### ISSUE TITLE
Unsupported platform for n@2.1.8: wanted {"os":"!win32","arch":"any"} (current: {"os":"win32","arch":"x64"})

### ISSUE THREAD

**Issue by zawsx** [tj/n#491]
# Issue Template:

### Describe issue including what OS you are using
Install seems confused by the Win32 subsystem on my x64 OS? (W10 Pro x64)
npm is in use and working otherwise.
### Describe what version of N you have
2.1.8
### Describe how you installed N
npm install -g n
### Steps to reproduce issue _1.2.3_
npm install -g n
### Describe the results you received
npm ERR! code EBADPLATFORM
npm ERR! notsup Unsupported platform for n@2.1.8: wanted {"os":"!win32","arch":"any"} (current: {"os":"win32","arch":"x64"})
npm ERR! notsup Valid OS:    !win32
npm ERR! notsup Valid Arch:  any
npm ERR! notsup Actual OS:   win32
npm ERR! notsup Actual Arch: x64

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\w10\AppData\Roaming\npm-cache\_logs\2018-03-05T12_06_16_565Z-debug.log
### Describe the results you expected
Clean Install
### What version of npm you are using
5.7.1
### Additional information you deem important (e.g. issue happens only occasionally):
node v9.2.0
Tried both as user and admin 
#### (paste your output here)
0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Users\\w10\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'install',
1 verbose cli   '-g',
1 verbose cli   'n' ]
2 info using npm@5.7.1
3 info using node@v9.2.0
4 verbose npm-session 12c2943e14a2e0a7
5 silly install loadCurrentTree
6 silly install readGlobalPackageData
7 http fetch GET 200 https://registry.npmjs.org/n 194ms
8 silly pacote tag manifest for n@latest fetched in 253ms
9 silly install loadIdealTree
10 silly install cloneCurrentTreeToIdealTree
11 silly install loadShrinkwrap
12 silly install loadAllDepsIntoIdealTree
13 silly resolveWithNewModule n@2.1.8 checking installable status
14 silly currentTree npm
15 silly idealTree npm
15 silly idealTree `-- n@2.1.8
16 silly install generateActionsToTake
17 silly diffTrees action count 1
18 silly diffTrees add n@2.1.8
19 silly decomposeActions action count 8
20 silly decomposeActions fetch n@2.1.8
21 silly decomposeActions extract n@2.1.8
22 silly decomposeActions preinstall n@2.1.8
23 silly decomposeActions build n@2.1.8
24 silly decomposeActions install n@2.1.8
25 silly decomposeActions postinstall n@2.1.8
26 silly decomposeActions finalize n@2.1.8
27 silly decomposeActions refresh-package-json n@2.1.8
28 silly install executeActions
29 silly doSerial global-install 8
30 verbose correctMkdir C:\Users\w10\AppData\Roaming\npm-cache\_locks correctMkdir not in flight; initializing
31 verbose lock using C:\Users\w10\AppData\Roaming\npm-cache\_locks\staging-db89fdc7e2954f2d.lock for C:\Users\w10\AppData\Roaming\npm\node_modules\.staging
32 silly doParallel extract 1
33 silly extract n@2.1.8
34 silly pacote trying n@latest by hash: sha512-VN3To6jBHSdTHmv+fT00V1nBdj3MWdcOlEKBEaUtXKRqQhuN289B9Jym9P/L3rXfafAsapmh4Fcr/DUR8DGJVg==
35 silly pacote data for sha512-VN3To6jBHSdTHmv+fT00V1nBdj3MWdcOlEKBEaUtXKRqQhuN289B9Jym9P/L3rXfafAsapmh4Fcr/DUR8DGJVg== not present. Using manifest.
36 http fetch GET 200 https://registry.npmjs.org/n/-/n-2.1.8.tgz 30ms
37 silly pacote n@latest extracted in 36ms
38 silly doReverseSerial unbuild 8
39 silly doSerial remove 8
40 silly doSerial move 8
41 silly doSerial finalize 8
42 silly finalize C:\Users\w10\AppData\Roaming\npm\node_modules\n
43 silly doParallel refresh-package-json 1
44 silly refresh-package-json C:\Users\w10\AppData\Roaming\npm\node_modules\n
45 verbose unlock done using C:\Users\w10\AppData\Roaming\npm-cache\_locks\staging-db89fdc7e2954f2d.lock for C:\Users\w10\AppData\Roaming\npm\node_modules\.staging
46 verbose stack Error: Unsupported platform for n@2.1.8: wanted {"name":"n","description":"Interactively Manage All Your Node Versions","version":"2.1.8","author":{"name":"TJ Holowaychuk","email":"tj@vision-media.ca"},"homepage":"https://github.com/tj/n","bugs":{"url":"https://github.com/tj/n/issues"},"contributors":[{"name":"Travis Webb","email":"me@traviswebb.com","url":"tjw.io"},{"name":"Nimit Kalra","email":"me@nimit.io","url":"http://nimit.io"},{"name":"Troy Connor","email":"troy0820@gmail.com","url":"https://github.com/troy0820"}],"keywords":["nvm","node","version","manager","switcher","node","binary","env"],"bin":{"n":"./bin/n"},"repository":{"type":"git","url":"git://github.com/tj/n.git"},"preferGlobal":true,"os":["!win32"],"engines":{"node":"*"},"license":"MIT","_resolved":"https://registry.npmjs.org/n/-/n-2.1.8.tgz","_integrity":"sha512-VN3To6jBHSdTHmv+fT00V1nBdj3MWdcOlEKBEaUtXKRqQhuN289B9Jym9P/L3rXfafAsapmh4Fcr/DUR8DGJVg==","_from":"n","_id":"n@2.1.8","bundleDependencies":false,"deprecated":false,"_shasum":"9b770f6ae0a93c49c64fb614625a63cc5a22d1fd","_requested":{"type":"tag","registry":true,"raw":"n","name":"n","escapedName":"n","rawSpec":"","saveSpec":null,"fetchSpec":"latest"},"_spec":"n","_where":"C:\\Users\\w10\\Documents\\cfg\\myscripts\\bin"} (current: {"os":"win32","cpu":"x64"})
46 verbose stack     at checkPlatform (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\npm-install-checks\index.js:45:14)
46 verbose stack     at tryCatcher (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\util.js:16:23)
46 verbose stack     at ret (eval at makeNodePromisifiedEval (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promisify.js:184:12), <anonymous>:13:39)
46 verbose stack     at readJson.then.catch.then (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\lib\install\action\refresh-package-json.js:31:12)
46 verbose stack     at tryCatcher (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\util.js:16:23)
46 verbose stack     at Promise._settlePromiseFromHandler (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:512:31)
46 verbose stack     at Promise._settlePromise (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:569:18)
46 verbose stack     at Promise._settlePromise0 (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:614:10)
46 verbose stack     at Promise._settlePromises (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:693:18)
46 verbose stack     at Promise._fulfill (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:638:18)
46 verbose stack     at Promise._settlePromise (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:582:21)
46 verbose stack     at Promise._settlePromise0 (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:614:10)
46 verbose stack     at Promise._settlePromises (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:693:18)
46 verbose stack     at Promise._fulfill (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:638:18)
46 verbose stack     at Promise._resolveCallback (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:432:57)
46 verbose stack     at Promise._settlePromiseFromHandler (C:\Users\w10\AppData\Roaming\npm\node_modules\npm\node_modules\bluebird\js\release\promise.js:524:17)
47 verbose pkgid n@2.1.8
48 verbose cwd C:\Users\w10\Documents\cfg\myscripts\bin
49 verbose Windows_NT 10.0.16299
50 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\w10\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "install" "-g" "n"
51 verbose node v9.2.0
52 verbose npm  v5.7.1
53 error code EBADPLATFORM
54 error notsup Unsupported platform for n@2.1.8: wanted {"os":"!win32","arch":"any"} (current: {"os":"win32","arch":"x64"})
55 error notsup Valid OS:    !win32
55 error notsup Valid Arch:  any
55 error notsup Actual OS:   win32
55 error notsup Actual Arch: x64
56 verbose exit [ 1, true ]

**Comment by jhr007** [comment]
Duplicate: https://github.com/tj/n/issues/435
tl;dr 
`n` doesn't support windows, or at least not x64 but brodybits on that issue mentioned https://github.com/jasongin/nvs

**Comment by zawsx** [comment]
Thanks for the response, appreciated! Wish I'd realised that Windows wasn't supported - maybe make the error message clearer?

**Comment by j2l** [comment]
And still not fixed, or even adding a clearer message in mid 2019.

**Comment by shadowspawn** [comment]
For the interest of future readers of this closed bug, and for @j2l.

`n` is written as a bash script. On plain Windows there is no bash shell, so `n` can not run. What we do is specify in our npm `package.json` that we do not support Windows, and the error and error message are handled by `npm` itself. We don't have any control over the message.

**Comment by danowar2k** [comment]
Could be important: On Windows there is the possibility to execute bash scripts using git for Windows bash or cygwin.

But perhaps there are other reasons n isn't usable on Windows...

**Comment by kthorne1** [comment]
And exactly how do I find where this is being requested or how can I exclude it. I don't see n@6.1.3 in my package.json nor when I search the entire project. But I'm getting this error on a Windows 10 box and need to resolve it. Thanks

**Comment by shadowspawn** [comment]
Hi @kthorne1 , you can list dependencies of a package using `npm ls`, and to see usage of `n`:
```
npm ls n
```

**Comment by csaar95** [comment]
I got exactly the same error, by running `npm install -g n`. It worked for me to update node with the .msi-installer on [](https://nodejs.org/en/).

**Comment by corg2001** [comment]
> For the interest of future readers of this closed bug, and for @j2l.
> 
> `n` is written as a bash script. On plain Windows there is no bash shell, so `n` can not run. What we do is specify in our npm `package.json` that we do not support Windows, and the error and error message are handled by `npm` itself. We don't have any control over the message.

How do you specify in npm 'package.json' that we do not support windows?

**Comment by shadowspawn** [comment]
> How do you specify in npm 'package.json' that we do not support windows?

https://github.com/tj/n/blob/158e5a07b7216589a0ea7dbce07f038ce25d13a7/package.json#L44

**Comment by corg2001** [comment]
@shadowspawn Thank you so much for your response, the n problem is not longer showing but still showing "Unsupported platform for <app-name>: wanted {"os":"!win32"} (current: {"os":"win32","arch":"x64"})

**Comment by shadowspawn** [comment]
`n` does not work with "Git for Windows bash", and an `npm` install of `n` gives the same somewhat cryptic error message from the original post as the operating system is detected as "win32".

`n` does potentially work in WSL running a compatible version of Linux, such as Ubuntu. See #511

**Comment by KonradLinkowski** [comment]
> `n` does not work with "Git for Windows bash", and an `npm` install of `n` gives the same somewhat cryptic error message from the original post as the operating system is detected as "win32".

Ok, but why doesn't it work? Is there any specific reason for that? Will it work one day?

**Comment by shadowspawn** [comment]
> Ok, but why doesn't it work? Is there any specific reason for that? Will it work one day?

`n` is a bash script. What `n` does is download and unpack a prebuilt tar archive of Node.js, and copy the files into place. The `node` executable in that archive has been build for Linux, requiring certain libraries and functionality, and won't run on Windows. So although `n` might be able to unpack the archive and copy the files, that version of node will not run.

There is a different prebuilt installer of Node.js for Windows. Running that installer is a different task than unpacking a tar archive.

WSL supports running the node binary built for Linux, and `n` works in WSL.
